### PR TITLE
feat(sdk): add runtime handling for custom artifacts (support custom artifact types pt. 2)

### PR DESCRIPTION
**Pull request by connor-mccarthy**
**Description of your changes:**
Adds runtime handling to support use of custom artifact types from non-kfp libraries in lightweight components. This includes (a) inferring import statement and injecting into `ephemeral_component.py` at runtime and (b) constructing a custom artifact type instance and passing it into the component.

For example, the following will no longer raise an exception at runtime:

```python
from kfp import dsl
from my_library.my_module import CustomType

@dsl.component
def my_component(custom_type: Output[CustomType]):
    custom_type.custom_method()
```

**Checklist:**
- [x] The title for your pull request (PR) should follow our title 
convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/compiler/test_data/pipelines/pipeline_with_custom_artifact_type.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+
+def create_temporary_non_kfp_artifact_package(
+        temp_dir: tempfile.TemporaryDirectory) -> None:
+    import inspect
+    import os
+    import textwrap
+
+    class ThirdPartyModel:
+        schema_title = 'third_party.ThirdPartyModel'
```
Comment (chensun) [inline sdk/python/kfp/compiler/test_data/pipelines/pipeline_with_custom_artifact_type.py:25]: This won't work. Schema title must reference some known artifact type that's pre-registered with MLMD backend. 
We need to stick with existing google namespaced types.

Comment (connor-mccarthy) [reply sdk/python/kfp/compiler/test_data/pipelines/pipeline_with_custom_artifact_type.py:25]: Agreed. This was included to test the compilation logic, not execution logic. There are no non-KFP libraries that have custom artifact classes that are compliant with KFP v2 artifacts.

Do you think adding a comment to indicate that this is a compilation test, not an execution test/example is sufficient until a non-KFP library adds compliant types?

Comment (chensun) [reply sdk/python/kfp/compiler/test_data/pipelines/pipeline_with_custom_artifact_type.py:27]: The compilation result shows `system.Artifact` which is incorrect if `third_party.ThirdPartyModel` were registered. Why not just use some existing google type, which make this sample executable today? Are you concerned about tinting the sample with some platform-specific elements?

Comment (connor-mccarthy) [reply sdk/python/kfp/compiler/test_data/pipelines/pipeline_with_custom_artifact_type.py:27]: I've updated the compilation logic -- waiting for tests to pass.

Is there currently a non-kfp library that implements a custom artifact type that is compatible with KFP artifacts and can be installed in the same environment as kfp v2? I do know of one, but would readily include one if it exists. I am not concerned about tinting the sample.

**Comment by connor-mccarthy** [comment]
Closing in favor of a more staged approach. See #8232 for next stage of implementation.

**Pull request closed without merging**